compiler/src/iree/compiler/ExternalInterfaces/EncodingExternalModels.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/ExternalInterfaces/EncodingExternalModels.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingDialect.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingOps.h"
+#include "iree/compiler/Dialect/Encoding/IR/EncodingTypes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Builders.h"
+
+namespace mlir::iree_compiler {
+namespace {
+
+struct ContractionAttrPropagationInterface
+    : public IREE::Encoding::EncodingPropagationAttrInterface::ExternalModel<
+          ContractionAttrPropagationInterface, IREE::Encoding::MatmulKAttr> {
+  bool isPropagable(Attribute attr, Value target) const {
+    auto encoding = dyn_cast<IREE::Encoding::MatmulKAttr>(attr);
+    if (!encoding) {
+      return false;
+    }
+    if (auto collapseOp = target.getDefiningOp<tensor::CollapseShapeOp>()) {
```

Comment:
I think it is better to use TypeSwitch, because we will add more ops to the list. The default can return `false`.

```cpp
return TypeSwitch<Operation*, bool> ...
```